cmd/pebble/cmd_logs.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+const (
+	logTimeFormat = "2006-01-02T15:04:05.000Z07:00"
+)
+
+type cmdLogs struct {
+	clientMixin
+	Follow     bool   `short:"f" long:"follow"`
+	Output     string `short:"o" long:"output"`
+	NumLogs    string `short:"n" long:"number"`
```

Comment:
I've looked around a bit at other tools, and everyone seems to use `-n` for this (see below for some examples). `journalctl` is perhaps the most similar tool, and it uses `-f`, `-n`, and `-o` like I've done. In particular `-n` and `-f` (follow) are common across all these tools. So I strongly suggest we stick to `-n` and `-f`.

I'm not so sold on `-o` / `--output` as I kind of agree with you it sounds like an output file name (that said, `journalctl` uses exactly that). Juju uses `--format` but doesn't allow the `-f` short format. In my mind `--format` seems very common.

My concrete suggestions:

* We keep `-f` / `--follow` as is.
* We keep `-n` but rename the long arg to `--num` and the field to `Num` (alternatively `--number` / `Number`).
* We drop `-o` and change `--output` to `--format` (no short form available).

Here is the (truncated) help output for the various commands I looked at:

```
journalctl
  -c --cursor=CURSOR       Show entries starting at the specified cursor
  -f --follow              Follow the journal
  -n --lines[=INTEGER]     Number of journal entries to show
  -o --output=STRING       Change journal output mode (short, short-iso,
                                   short-precise, short-monotonic, verbose,
                                   export, json, json-pretty, json-sse, cat)

tail
  -c, --bytes=[+]NUM       output the last NUM bytes; or use -c +NUM to
                             output starting with byte NUM of each file
  -f, --follow[={name|descriptor}]
                           output appended data as the file grows;
                             an absent option argument means 'descriptor'
  -n, --lines=[+]NUM       output the last NUM lines, instead of the last 10;

juju debug-log  # follow is the default for this, --no-tail turns it off
-n, --lines  (= 10)
    Show this many of the most recent (possibly filtered) lines, and continue to append

snap logs
      -n=              Show only the given number of lines, or 'all'. (default: 10)
      -f               Wait for new lines and print them as they come in.

docker logs
  -f, --follow         Follow log output
  -n, --tail string    Number of lines to show from the end of the logs (default "all")

kubectl logs
  -c, --container='': Print the logs of this container
  -f, --follow=false: Specify if the logs should be streamed.
      --tail=-1: Lines of recent log file to display. Defaults to -1 with no selector, showing all log lines otherwise 10, if a selector is provided.
```

Reply:
Seeing we're doing this async, I've proactively made the changes I suggested above, but it's easy to tweak as needed based on further discussion.

Reply:
Thanks, it looks like a good direction. Can we please just do one more tweak on top of it: let's drop the long form (`--num` and `Num` now) altogether and have `-n` and `.N` only. I think that's also what we do on the API already, isn't it?  So we're standardizing across all cases with the most expectable name, and it's a nice one as well.

Reply:
Good call. Yep, "n" is what we do in the API too.